providers/implementations/rands/fips_crng_test.c

```diff
@@ -0,0 +1,428 @@
+{
+    return 1;
+}
+
+static size_t crng_test_get_seed(void *vcrngt, unsigned char **pout,
+                                 int entropy, size_t min_len,
+                                 size_t max_len,
+                                 int prediction_resistance,
+                                 const unsigned char *adin,
+                                 size_t adin_len)
+{
+    CRNG_TEST *crngt = (CRNG_TEST *)vcrngt;
+    size_t n;
+    int r = 0;
+
+    /* Without a parent, we rely on the up calls */
+    if (crngt->parent == NULL
+            || crngt->parent_get_seed == NULL) {
+        n = ossl_prov_get_entropy(crngt->provctx, pout, 0, min_len, max_len);
+        if (n == 0)
+            return 0;
+        r = crng_test(crngt, *pout, n);
+        return r > 0 ? n : 0;
+    }
+
+    /* Grab seed from our parent */
+    if (!crng_test_lock(crngt))
+        return 0;
+    if (!lock_parent(crngt)) {
+        crng_test_unlock(crngt);
+        return 0;
+    }
+    n = crngt->parent_get_seed(crngt->parent, pout, entropy,
+                               min_len, max_len, prediction_resistance,
+                               adin, adin_len);
+    unlock_parent(crngt);
+    if (n > 0)
+        r = crng_test(crngt, *pout, n);
+    crng_test_unlock(crngt);
+    if (n > 0 && r > 0)
+        return n;
+    if (crngt->parent_clear_seed != NULL)
+        crngt->parent_clear_seed(crngt->parent, *pout, n);
+    return 0;
+}
+
+static void crng_test_clear_seed(void *vcrngt,
+                                 unsigned char *out, size_t outlen)
+{
+    CRNG_TEST *crngt = (CRNG_TEST *)vcrngt;
+
+    if (crngt->parent == NULL || crngt->parent_get_seed == NULL)
+        ossl_prov_cleanup_entropy(crngt->provctx, out, outlen);
+    else if (crngt->parent_clear_seed != NULL)
+        crngt->parent_clear_seed(crngt->parent, out, outlen);
+}
+
+static int crng_test_enable_locking(void *vcrngt)
+{
+    CRNG_TEST *crngt = (CRNG_TEST *)vcrngt;
+
+    if (crngt != NULL && crngt->lock == NULL) {
+        if (crngt->parent_enable_locking != NULL)
+            if (!crngt->parent_enable_locking(crngt->parent)) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_PARENT_LOCKING_NOT_ENABLED);
+                return 0;
+            }
+        crngt->lock = CRYPTO_THREAD_lock_new();
+        if (crngt->lock == NULL) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_CREATE_LOCK);
+            return 0;
+        }
+    }
+    return 1;
+}
+
+static int crng_test_lock(ossl_unused void *vcrngt)
+{
+    CRNG_TEST *crngt = (CRNG_TEST *)vcrngt;
+
+    return crngt->lock == NULL || CRYPTO_THREAD_write_lock(crngt->lock);
+}
+
+static void crng_test_unlock(ossl_unused void *vcrngt)
+{
+    CRNG_TEST *crngt = (CRNG_TEST *)vcrngt;
+
+    if (crngt->lock != NULL)
+        CRYPTO_THREAD_unlock(crngt->lock);
+}
+
+static int crng_test_get_ctx_params(void *vcrngt, OSSL_PARAM params[])
+{
+    CRNG_TEST *crngt = (CRNG_TEST *)vcrngt;
+    OSSL_PARAM *p;
+
```

Comment:
Should this have an indicator like test_rng_get_ctx_params() ??

#ifdef FIPS_MODULE
    p = OSSL_PARAM_locate(params, OSSL_RAND_PARAM_FIPS_APPROVED_INDICATOR);
    if (p != NULL && !OSSL_PARAM_set_int(p, 0))
        return 0;
#endif  /* FIPS_MODULE */

Reply:
I've added this, without the FIPS_MODULE guard since this is only built for the FIPS module and is not available in the default or base providers.